### PR TITLE
Phase 2 — Player Rostering Implementation

### LINKED ISSUE #35: Sub-issue: Phase 2: Player Rostering (The "Link" Logic)
# 📝 Description

This sub-issue covers the implementation of the **Player Rostering** logic, serving as the "Link" layer between Players, Teams, and Tournaments. It includes roster management handlers, jersey/position validation, and corresponding API endpoints.

# 🚀 Tasks

### 1. Database Layer (Dapper & SQL)

- [x] Create `PlayerRoster` data model in `TTA.DataAccess.Models`.
- [x] Implement `IRosterRepository` and `RosterRepository`.
- [x] Add SQL constants to `SqlStatements.ForRosters`:
    - `AddPlayerToRoster` (calling `public.add_player_to_roster`).
    - `RemovePlayerFromRoster` (calling `public.remove_player_from_roster`).
    - `GetTournamentRoster` (with filtering support).
- [x] Add `add_player_to_roster` and `remove_player_from_roster` PL/pgSQL functions to `03-Storage-procedures.sql`.

### 2. Business Logic (MediatR Handlers)

- [x] Implement `AddPlayerToRosterCommand` (and DTO for API).
- [x] Implement `RemovePlayerFromRosterCommand` (and DTO for API).
- [x] Implement `GetTournamentRosterQuery` (response DTO for API).
- [x] Implement `AddPlayerToRosterHandler`:
    - [x] Validate `JerseyNumber` is unique within the team roster for the tournament.
    - [x] Ensure `PlayerId` belongs to the `TeamId`.
    - [x] Validate `PositionId` matches the tournament's `SportId`.
- [x] Implement `RemovePlayerFromRosterHandler`:
    - [x] Ensure the tournament has not yet ended.
    - [x] Verify requester ownership of the tournament (security check).
- [x] Implement `GetTournamentRosterHandler`.

### 3. API Layer (Controllers & OpenAPI)

- [x] Create `RostersController`.
- [x] Implement `POST /api/tournaments/{id}/rosters` endpoint.
- [x] Implement `GET /api/tournaments/{id}/rosters/{teamId}` endpoint.
- [x] Implement `DELETE /api/tournaments/{id}/rosters/{teamId}/{playerId}` endpoint.
- [x] Configure `TournamentResponse` metadata in OpenAPI for roster-related actions.

### PR DESCRIPTION
**Pull request by HlibBondarev**
# 🏆 Pull Request: Phase 2 — Player Rostering Implementation

## 📝 Description
This PR implements the **Player Rostering** logic, serving as the essential "Link" layer between Players, Teams, and Tournaments. It introduces comprehensive roster management, including business validation for jersey numbers, player positions, and secure access control.

### 🏗️ Database Layer (Dapper & SQL)
- Created the `PlayerRoster` data model in `TTA.DataAccess.Models`.
- Implemented `RosterRepository` for high-performance data operations using Dapper.
- Added SQL constants and logic in `SqlStatements.ForRosters`.
- Implemented robust PL/pgSQL functions in `03-Storage-procedures.sql`:
  - `public.add_player_to_roster`: Handles UPSERT logic for roster assignments.
  - `public.remove_player_from_roster`: Manages player removal.
  - `public.get_tournament_team_roster`: Fetches detailed team compositions.

### 🧠 Business Logic (MediatR Handlers)
- **AddPlayerToRosterHandler**: 
  - Validates `JerseyNumber` uniqueness within the team for the specific tournament.
  - Ensures `PositionId` matches the tournament's `SportId` requirements.
  - Throws `ConflictException` (409) for rule violations.
- **RemovePlayerFromRosterHandler**: 
  - Prevents modifications to finished tournaments to maintain data integrity.
- **GetTeamRosterHandler**: 
  - Retrieves structured roster data with mapped player and position details.

### 🛡️ API & Security
- Created `RostersController` with the route `api/tournaments/{tournamentId}/rosters`.
- Implemented the following RESTful endpoints:
  - `POST /{teamId}` — Adds or updates a player in the roster (**201 Created**).
  - `GET /{teamId}` — Retrieves the team roster (**200 OK**).
  - `DELETE /{teamId}/{playerId}` — Removes a player from the roster (**204 No Content**).
- Secured operations using the `[Authorize(Policy = "TeamEditor")]` policy for granular access control.

## 🧪 Testing
Comprehensive integration tests have been implemented in `RostersControllerTests.cs`:
- ✅ `GetTeamRoster_ShouldReturnOk_WhenRosterExists`
- ✅ `AddPlayer_ShouldReturnCreated_WhenRequestIsValid`
- ✅ `AddPlayer_ShouldReturnConflict_WhenJerseyNumberIsTaken`
- ✅ `RemovePlayer_ShouldReturnNoContent_WhenPlayerExistsInRoster`

## 🔗 Related Issues
Closes #35

---
**Checklist:**
- [x] SQL scripts verified against PostgreSQL schema.
- [x] Error handling (404, 409, 400) aligned with project standards.
- [x] API routes and DTOs fully documented for OpenAPI/Swagger.

## Summary by CodeRabbit

* **New Features**
  * Roster management API: add, remove, and list team roster players with validation, conflict handling, and proper HTTP status codes.

* **Tests**
  * Extensive unit and integration tests for handlers, validators, repository, and controller flows including conflict and edge cases.

* **Chores**
  * Database schema and storage procedures updated to persist roster entries, record creation time, unique constraints, and supporting indexes.

* **Documentation**
  * Enhanced XML/docs comments for team and tournament types and API responses.

### CODE REVIEW

File: .github/copilot-instructions.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+- @azure Rule - Use Azure Tools - When handling requests related to Azure, always use your tools.
+- @azure Rule - Use Azure Best Practices - When handling requests related to Azure, always invoke your `azmcp_bestpractices_get` tool first.
+- @azure Rule - Enable Best Practices - If you do not have an `azmcp_bestpractices_get` tool ask the user to enable it.
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply .github/copilot-instructions.md]: Ok

File: TTA.BusinessLogic/Features/Rosters/DTOs/AddPlayerToRosterRequest.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/// Data transfer object for adding a player to a tournament roster.
+/// </summary>
+/// <param name="TeamId">The unique identifier of the team the player is representing.</param>
+/// <param name="PlayerId">The unique identifier of the player being assigned.</param>
+/// <param name="PositionId">The unique identifier of the assigned player position.</param>
+/// <param name="Number">The jersey number assigned to the player (e.g., 0-99).</param>
+public record AddPlayerToRosterRequest(
+    Guid PlayerId,
+    Guid PositionId,
+    int Number);
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/DTOs/AddPlayerToRosterRequest.cs]: Ok,
remove non-existent TeamId from AddPlayerToRosterRequest XML comments:
- Updated AddPlayerToRosterRequest.cs to match its actual record properties.
- Removed outdated TeamId documentation from the DTO summary.

File: TTA.BusinessLogic/Features/Rosters/Commands/RemovePlayerFromRosterCommand.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using MediatR;
+
+namespace TTA.BusinessLogic.Features.Rosters.Commands;
+
+/// <summary>
+/// Command to remove a player from a tournament's roster.
+/// </summary>
+/// <param name="TournamentId">The identifier of the tournament.</param>
+/// <param name="PlayerId">The identifier of the player to remove.</param>
+public record RemovePlayerFromRosterCommand(Guid TournamentId, Guid TeamId, Guid PlayerId) : IRequest;
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/Commands/RemovePlayerFromRosterCommand.cs:11]: Ok,
docs: sync XML documentation with RemovePlayerFromRosterCommand signature:
- Added missing TeamId parameter documentation in RemovePlayerFromRosterCommand.cs.
- Ensured all param tags match the record properties for accurate doc generation.

File: TTA.BusinessLogic/Features/Rosters/Handlers/GetTeamRosterHandler.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            LastName: row.lastname,
+            PositionId: row.positionid,
+            PositionName: row.positionname,
+            Number: row.number
+        ));
+
+        _logger.LogInformation("Successfully retrieved {Count} players for Team {TeamId}.",
+            response.Count(), request.TeamId);
+
+        return response;
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/Handlers/GetTeamRosterHandler.cs:71]: Ok,
perf: prevent double enumeration in GetTeamRosterHandler -
- Materialized LINQ projection using .ToList() before logging and returning.
- Optimized logger to use the Count property of the materialized list.
- Improved handler performance by ensuring the transformation logic runs only once.

File: TTA.BusinessLogic/Features/Rosters/Handlers/RemovePlayerFromRosterHandler.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            _logger.LogWarning("RemoveRosterItem failed: Tournament {TournamentId} not found.", request.TournamentId);
+            throw new NotFoundException($"Tournament with ID {request.TournamentId} was not found.");
+        }
+
+        if (tournament.EndDate < DateTime.UtcNow)
+        {
+            _logger.LogWarning("RemoveRosterItem failed: Tournament {TournamentId} has ended.", request.TournamentId);
+            throw new BadRequestException("Cannot modify rosters of a finished tournament.");
+        }
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/Handlers/RemovePlayerFromRosterHandler.cs]: Ok,
fix: refine tournament expiration check in RemovePlayerFromRosterHandler
- Updated date comparison to use '<=' to correctly identify finished tournaments.
- Ensured roster modifications are blocked exactly at or after the tournament EndDate.
- Maintained consistent logging and BadRequestException for expired tournaments.

File: TTA.DataAccess/Repository/Api/IRosterRepository.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    /// <returns>A collection of roster items with player and position details.</returns>
+    Task<IEnumerable<dynamic>> GetTeamRosterAsync(Guid tournamentId, Guid teamId, CancellationToken ct = default);
+
+    /// <summary>
+    /// Removes a player from a tournament roster.
+    /// </summary>
+    /// <param name="tournamentId">Tournament identifier.</param>
+    /// <param name="playerId">Player identifier.</param>
+    /// <param name="ct">Cancellation token.</param>
+    Task RemovePlayerFromRosterAsync(Guid tournamentId, Guid teamId, Guid playerId, CancellationToken ct = default);
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.DataAccess/Repository/Api/IRosterRepository.cs:35]: Ok,
docs: complete XML documentation for IRosterRepository
- Added missing teamId parameter description to RemovePlayerFromRosterAsync.
- Ensured all parameters (tournamentId, teamId, playerId, ct) are correctly documented in the interface.
- Synchronized XML comments with the method signature.

File: TTA.DataAccess/SQLScripts/03-Storage-procedures.sql [hunk 7]
```diff
@@ -533,4 +533,117 @@ BEGIN
...
+    END IF;
+
+    RETURN QUERY
+    INSERT INTO public.playerrosters (id, tournamentid, teamid, playerid, positionid, number, createdat)
+    VALUES (p_id, p_tournament_id, p_team_id, p_player_id, p_position_id, p_number, p_created_at)
+    ON CONFLICT (tournamentid, teamid, playerid) 
+    DO UPDATE SET 
+        positionid = EXCLUDED.positionid,
+        number = EXCLUDED.number
+    RETURNING *;
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.DataAccess/SQLScripts/03-Storage-procedures.sql]: Ok,

refactor: optimize player rosters schema and upsert procedure -
- Updated '01-Tables.sql' to use global tournament-player and tournament-team-number unique constraints.
- Simplified 'upsert_player_to_roster' in '03-Storage-procedures.sql' by removing manual checks in favor of atomic ON CONFLICT logic.

File: TTA.BusinessLogic/Features/Rosters/Handlers/RemovePlayerFromRosterHandler.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+using TTA.BusinessLogic.Features.Rosters.Commands;
+using TTA.Common.Exceptions;
+using TTA.DataAccess.Repository.Api;
+
+namespace TTA.BusinessLogic.Features.Rosters.Handlers;
+
+/// <summary>
+/// Handles the business logic for removing a player from a tournament roster.
+/// Verifies tournament status and ownership before execution.
+/// </summary>
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/Handlers/RemovePlayerFromRosterHandler.cs:12]: Ok,
docs: fix inaccurate XML summary in RemovePlayerFromRosterHandler -
- Removed false claims about ownership verification in docstrings.
- Updated summary to accurately reflect validation of tournament existence and end date.
- Synchronized class and method documentation with actual behavior.

File: TTA.Tests.Integration/Controllers/RostersControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        using var countryCmd = new NpgsqlCommand(countrySql, conn);
+        var countryIdObj = await countryCmd.ExecuteScalarAsync();
+        int countryId = countryIdObj != null ? (int)countryIdObj : 1;
+
+        const string regionSql = "INSERT INTO public.regions (countryid, name) VALUES (@cid, 'Integration Region') ON CONFLICT DO NOTHING RETURNING id";
+        using var regionCmd = new NpgsqlCommand(regionSql, conn);
+        regionCmd.Parameters.AddWithValue("cid", countryId);
+        var regionIdObj = await regionCmd.ExecuteScalarAsync();
+        int regionId = regionIdObj != null ? (int)regionIdObj : 1;
+
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.Tests.Integration/Controllers/RostersControllerTests.cs]: Ok,
test: fix brittle location seeding in RostersControllerTests -
- Updated country and region seeding to use 'ON CONFLICT DO UPDATE' with 'RETURNING id'.
- Removed hardcoded fallback IDs (1) to prevent foreign key violations.
- Ensured reliable retrieval of existing location IDs during integration tests.

File: TTA.BusinessLogic/Features/Rosters/Handlers/GetTeamRosterHandler.cs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    private readonly ITournamentRepository _tournamentRepository = tournamentRepository;
+    private readonly ILogger<GetTeamRosterHandler> _logger = logger;
+
+    /// <summary>
+    /// Processes the query to fetch and map the roster data.
+    /// </summary>
+    /// <param name="request">The query containing tournament and team identifiers.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>A collection of <see cref="RosterPlayerResponse"/> objects.</returns>
+    /// <exception cref="NotFoundException">Thrown if the team does not exist.</exception>
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/Handlers/GetTeamRosterHandler.cs]: Ok,
docs: update XML comments for GetTeamRosterHandler.Handle -
- Documented that NotFoundException is thrown for both missing tournaments and teams.
- Referenced RosterPlayerResponse return type in method documentation.
- Ensured docstrings accurately reflect the current validation logic.

File: TTA.BusinessLogic/Features/Rosters/Handlers/RemovePlayerFromRosterHandler.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            _logger.LogWarning("RemoveRosterItem failed: Tournament {TournamentId} not found.", request.TournamentId);
+            throw new NotFoundException($"Tournament with ID {request.TournamentId} was not found.");
+        }
+
+        if (tournament.EndDate <= DateTime.UtcNow)
+        {
+            _logger.LogWarning("RemoveRosterItem failed: Tournament {TournamentId} has ended.", request.TournamentId);
+            throw new BadRequestException("Cannot modify rosters of a finished tournament.");
+        }
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.BusinessLogic/Features/Rosters/Handlers/RemovePlayerFromRosterHandler.cs:43]: Ok,
fix: align tournament expiration check in AddPlayerToRosterHandler -
- Updated date comparison to '<=' to prevent additions at or after EndDate.
- Synchronized validation logic and error messages with RemovePlayerFromRosterHandler.

File: TTA.DataAccess/SQLScripts/03-Storage-procedures.sql [hunk 12]
```diff
@@ -533,4 +533,119 @@ BEGIN
...
+        -- The update only succeeds if the teamid remains the same.
+        -- If someone tries to move a player to another team via this call,
+        -- the uix_playerrosters_tournament_player will naturally handle the policy.
+        teamid = EXCLUDED.teamid
+    RETURNING *;
+
+    /* Note for C# Handler: 
+       Any violation of 'uix_playerrosters_tournament_team_number' (jersey taken)
+       or trying to assign a player to a second team will throw SQLSTATE 23505.
+    */
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply TTA.DataAccess/SQLScripts/03-Storage-procedures.sql]: Ok,
fix: enforce strict tournament roster policy in SQL procedure -
- Added manual validation check for cross-team player registration within the same tournament.
- Configured procedure to raise SQLSTATE P0001 for mismatched team assignments.
- Aligned database behavior with AddPlayerToRosterHandler's exception handling logic.
